File: mathlib/derivations/simplifications/MultiplyExponentPowers.ts
import { Argument } from "../../Argument";
import { product } from "../../ConvenientExpressions";
import { Exponent } from "../../expressions/Exponent";
import { Expression } from "../../expressions/Expression";
import { Relationship } from "../../Relationship";
import { setOf } from "../../util/ThingsThatShouldBeInTheStdLib";
import { NoContextExpressionSimplificationRule } from "../NoContextExpressionSimplificationRule";

export class MultiplyExponentPowers extends NoContextExpressionSimplificationRule {
    protected appliesImpl(exp: Expression): boolean {
        return exp instanceof Exponent
            && exp.base instanceof Exponent

    }
    protected applyImpl(exp: Expression): Set<Argument> {
        const outer = exp as Exponent
        const inner = outer.base as Exponent

        return setOf(new Argument(setOf(exp), {
            n: exp,
            r: Relationship.Equal,
            n1: Exponent.of(inner.base, product(inner.power, outer.power))
        }, "Exponents to exponents multiply"))
    }

}

File: cas/src/derivation_rules/multiply_exponent_powers.rs
use std::rc::Rc;

use crate::{expressions::{ExpressionPtr, Expression, product::product_of}, argument::Argument, convenience_expressions::power};

use super::DerivationRule;


/**
* a^b^c = a^(bc)
*/
pub struct MultiplyExponentPowers {}

impl DerivationRule for MultiplyExponentPowers {
    fn apply(&self, input: ExpressionPtr) -> Vec<(ExpressionPtr, Rc<Argument>)> {
        let outer =  match input {
            Expression::Exponent(ref e) => e,
            _ => return vec![],
        };

        let inner = match outer.power() {
            Expression::Exponent(e) => e,
            _ => return vec![],
        };

        vec![(power(outer.base(), product_of(&[inner.base(), inner.power()])),
            Argument::new("Nested exponents multiply".to_owned(), vec![input]))]
    }
}

#[cfg(test)]
mod tests {
    use crate::{expressions::product::product_of, convenience_expressions::{power, i}};

    use super::MultiplyExponentPowers;
    use crate::derivation_rules::DerivationRule;

    #[test]
    fn test_1() {
        let start = power(i(1), power(i(2), i(3)));
        let rule = MultiplyExponentPowers{};
        let result = rule.apply(start);
        assert_eq!(result.first().unwrap().0, power(i(1), product_of(&[i(2), i(3)])));
    }
}
